net/netdev/netdev_ioctl.c

```diff
@@ -636,235 +636,204 @@ static FAR struct net_driver_s *netdev_ifr_dev(FAR struct ifreq *req)
 static int netdev_ifr_ioctl(FAR struct socket *psock, int cmd,
                             FAR struct ifreq *req)
 {
-  FAR struct net_driver_s *dev;
-  int ret = -EINVAL;
+  FAR struct net_driver_s *dev = NULL;
+  FAR struct net_driver_s *tmpdev;
+  ssize_t arglen;
+  int ret;
 
   ninfo("cmd: %d\n", cmd);
 
   net_lock();
 
-  /* Execute the command */
+  /* Execute the command without ifr_name or lifr_name */
 
   switch (cmd)
     {
+      case SIOCGIFCOUNT:  /* Get number of devices */
+        {
+          req->ifr_count = netdev_count();
+          ret = OK;
+        }
+        break;
+
 #ifdef CONFIG_NET_IPv4
-      case SIOCGIFADDR:  /* Get IP address */
+      case SIOCGIFCONF:  /* Return an interface list (IPv4) */
         {
-          dev = netdev_ifr_dev(req);
-          if (dev)
-            {
-              ioctl_get_ipv4addr(&req->ifr_addr, dev->d_ipaddr);
-              ret = OK;
-            }
+          ret = netdev_ipv4_ifconf((FAR struct ifconf *)req);
         }
         break;
 #endif
 
-#ifdef CONFIG_NET_IPv4
-      case SIOCSIFADDR:  /* Set IP address */
+#ifdef CONFIG_NET_IPv6
+      case SIOCGLIFCONF:  /* Return an interface list (IPv6) */
         {
-          dev = netdev_ifr_dev(req);
-          if (dev)
-            {
-              ioctl_set_ipv4addr(&dev->d_ipaddr, &req->ifr_addr);
-              ret = OK;
-            }
+          ret = netdev_ipv6_ifconf((FAR struct lifconf *)req);
         }
         break;
 #endif
 
-#ifdef CONFIG_NET_IPv4
-      case SIOCGIFDSTADDR:  /* Get P-to-P address */
+#ifdef CONFIG_NETDEV_IFINDEX
+      case SIOCGIFNAME:  /* Get interface name */
         {
-          dev = netdev_ifr_dev(req);
-          if (dev)
+          tmpdev = netdev_findbyindex(req->ifr_ifindex);
```

Comment:
why not use dev directly

Reply:
Because originally there was one `switch`, but now there are two.

The first `switch` handles IOCTLs that do not require `dev`, and the `default` gets the `dev`.

Then there is a check if `dev` is `NULL` then we return.

Then, the second `switch` handles all the other IOCTLs.

If use `dev` directly, then we will execute the second `switch` when we shouldn't.

This is a nice refactoring because:

1. Instead of many places getting `dev`, now there is only one. (Reduce code size.)
2. A lot of `#ifdef` rash is reduced by grouping IPv4, IPv6, Ethernet, and 6LOWPAN IOCTLs.

I think it is correct (just by code reading). The diff is hard to read, though, because diff algorithm is grouping things strangely.

Reply:
Hmm... compiler is complaining tmpdev is unused?

Reply:
Maybe should declare `tmpdev` inside the block where it is used? This will also avoid future mistakes of wrongfully using `tmpdev` instead of `dev` in other places.

Reply:
Done